src/grain/createGrainDistribution.js

```diff
@@ -0,0 +1,269 @@
+// @flow
+
+/**
+ * In SourceCred, projects regularly distribute Grain to contributors based on
+ * their Cred scores. This is called a "Distribution".
+ *
+ * This module contains the logic for calculating distribution amounts for
+ * contributors. A distribution contains "receipts" showing how much Grain each
+ * contributor will receive, the timestamp of the distribution in question, and the
+ * "strategy" used to distribute grain.
+ *
+ * Currently we support two strategies:
+ * - IMMEDIATE, which distributes a fixed budget of grain based on the cred scores
+ * in the most recent completed time interval
+ * - BALANCED, which distributes a fixed budget of grain based on cred scores
+ * across
+ * all time, prioritizing paying people who were under-paid historically (i.e.
+ * their lifetime earnings are lower than we would expect given their current
+ * cred score)
+ *
+ * In all cases, the timestamp for the distribution is used to determine which cred
+ * scores are in scope. For example, if you create a immediate distribution with a
+ * timestamp in the past, it will reward people with cred in the past time
+ * period, not the current one.
+ */
+import {sum} from "d3-array";
+import {mapToArray} from "../util/map";
+import {type NodeAddressT} from "../core/graph";
+import {
+  type Grain,
+  multiplyFloat,
+  ZERO,
+  toFloatRatio,
+  format,
+  DECIMAL_PRECISION,
+} from "./grain";
+
+export const GRAIN_DISTRIBUTION_VERSION_1 = 1;
+
+export type DistributionStrategy = ImmediateV1 | BalancedV1;
+
+export type ImmediateV1 = {|
+  +type: "IMMEDIATE",
+  +version: number,
+  +budget: Grain,
+|};
+
+export type BalancedV1 = {|
+  +type: "BALANCED",
+  +version: number,
+  +budget: Grain,
+|};
+
+export type GrainReceipt = {|
+  +address: NodeAddressT,
+  +amount: Grain,
+|};
+
+export type GrainDistributionV1 = {|
+  +type: "GRAIN_DISTRIBUTION",
+  +timestampMs: number,
+  +version: number,
+  +strategy: DistributionStrategy,
+  +receipts: $ReadOnlyArray<GrainReceipt>,
+|};
+
+export type CredTimeSlice = {|
+  +intervalEndMs: number,
+  +cred: Map<NodeAddressT, number>,
+|};
+
+export type CredHistory = $ReadOnlyArray<CredTimeSlice>;
+
+/**
+ * Compute a full Distribution given:
+ * - the strategy we're using
+ * - the full cred history for all users
+ * - the lifetime earnings of all users
```

Comment:
let's have a slightly longer comment here explaining what lifetime earnings means in terms of past grain distributions